hphp/runtime/vm/jit/vasm-arm.cpp

```diff
@@ -1328,55 +1275,145 @@ void lower(Vunit& unit, cvttsd2siq& i, Vlabel b, size_t idx) {
+  v << callr{scratch, i.args};
+  });
+}
+
+void lower(Vunit& u, jmpm& i, Vlabel b, size_t z) {
+  lower_impl(u, b, z, [&] (Vout& v) {
+  lowerVptr(i.target, v);
+
+  auto scratch = v.makeReg();
+
+  v << load{i.target, scratch};
+  v << jmpr{scratch, i.args};
+  });
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+void lower(Vunit& u, callstub& i, Vlabel b, size_t z) {
+  lower_impl(u, b, z, [&] (Vout& v) {
+    v << call{i.target, i.args};
+  });
+}
+
+void lower(Vunit& u, callfaststub& i, Vlabel b, size_t z) {
+  lower_impl(u, b, z, [&] (Vout& v) {
+    v << call{i.target, i.args};
+    v << syncpoint{i.fix};
+  });
+}
+
+void lower(Vunit& u, tailcallstub& i, Vlabel b, size_t z) {
+  lower_impl(u, b, z, [&] (Vout& v) {
+    // Restore LR from native stack and adjust SP
+    v << popp{PhysReg(rAsm), rlr()};
+
+    // Then directly jump to the target
+    v << jmpi{i.target, i.args};
+  });
+}
+
+void lower(Vunit& u, stubret& i, Vlabel b, size_t z) {
+  lower_impl(u, b, z, [&] (Vout& v) {
+    // Pop LR and (optionally) FP
+    if (i.saveframe) {
+      v << popp{rfp(), rlr()};
+    } else {
+      v << popp{PhysReg(rAsm), rlr()};
+    }
+
+    // Return
+    v << ret{i.args};
+  });
+}
+
+void lower(Vunit& u, stublogue& i, Vlabel b, size_t z) {
+  lower_impl(u, b, z, [&] (Vout& v) {
+    // Push both the LR and FP regardless of i.saveframe to align SP
+    v << pushp{rfp(), rlr()};
+  });
+}
+
 void lower(Vunit& u, stubunwind& i, Vlabel b, size_t z) {
   lower_impl(u, b, z, [&] (Vout& v) {
+    // Pop the call frame
     v << lea{rsp()[16], rsp()};
   });
 }
 
 void lower(Vunit& u, stubtophp& i, Vlabel b, size_t z) {
   lower_impl(u, b, z, [&] (Vout& v) {
+    // Pop the call frame
     v << lea{rsp()[16], rsp()};
   });
 }
 
 void lower(Vunit& u, loadstubret& i, Vlabel b, size_t z) {
   lower_impl(u, b, z, [&] (Vout& v) {
+    // Load the LR to the destination
     v << load{rsp()[8], i.d};
   });
 }
 
-void lower(Vunit& u, calltc& i, Vlabel b, size_t z) {
+///////////////////////////////////////////////////////////////////////////////
+
+void lower(Vunit& u, phplogue& i, Vlabel b, size_t z) {
   lower_impl(u, b, z, [&] (Vout& v) {
-    // Push FP, LR for callToExit(..)
-    auto r0 = v.makeReg();
-    auto r1 = v.makeReg();
-    v << load{i.fp[AROFF(m_savedRip)], r0};
-    v << ldimmq{i.exittc, r1};
-    v << pushp{r0, r1};
-
-    // Emit call to next instruction to balance predictor's stack
-    v << bln{};
-
-    // Set the return address to savedRip and jump to target
-    v << copy{r0, rlr()};
+    // Save the LR to the saved RIP
+    v << store{rlr(), i.fp[AROFF(m_savedRip)]};
+  });
+}
+
+void lower(Vunit& u, phpret& i, Vlabel b, size_t z) {
+  lower_impl(u, b, z, [&] (Vout& v) {
+    // Restore the saved RIP to LR
+    v << load{i.fp[AROFF(m_savedRip)], rlr()};
+
+    // If there's no frame, then load fp
+    if (!i.noframe) {
+      v << load{i.fp[AROFF(m_sfp)], i.d};
+    }
+
+    // Return
```

Comment:
These inline comments all seem a little redundant.  Try to avoid comments of the form

```
// Do the thing.
do_the_thing();
```